File: python-sidecar/pipeline/ssrf_guard.py
"""SSRF guard for sidecar fetches of user-supplied URLs."""
from __future__ import annotations

import ipaddress
import socket
from typing import Mapping
from urllib.parse import urljoin, urlparse

import httpx


def _is_private_address(address: str) -> bool:
    ip = ipaddress.ip_address(address)
    if (
        ip.is_private
        or ip.is_loopback
        or ip.is_link_local
        or ip.is_multicast
        or ip.is_reserved
        or ip.is_unspecified
    ):
        return True
    if ip.version == 4 and ipaddress.ip_address("100.64.0.0") <= ip <= ipaddress.ip_address("100.127.255.255"):
        return True
    return False


def assert_public_url(raw_url: str) -> None:
    parsed = urlparse(raw_url)
    if parsed.scheme not in {"http", "https"} or not parsed.hostname:
        raise ValueError("Only http(s) URLs are allowed")

    host = parsed.hostname.strip("[]").lower()
    if host == "localhost" or host.endswith(".localhost"):
        raise ValueError("Blocked host")

    try:
        if _is_private_address(host):
            raise ValueError("Blocked private address")
        return
    except ValueError as exc:
        if str(exc).startswith("Blocked"):
            raise

    try:
        infos = socket.getaddrinfo(host, None, type=socket.SOCK_STREAM)
    except socket.gaierror as exc:
        raise ValueError("DNS resolution failed") from exc

    addresses = {info[4][0] for info in infos}
    if not addresses:
        raise ValueError("DNS resolution failed")
    for address in addresses:
        if _is_private_address(address):
            raise ValueError("Blocked private address")


async def ssrf_safe_get(
    url: str,
    *,
    headers: Mapping[str, str] | None = None,
    timeout: float = 20,
    max_redirects: int = 5,
    verify: bool = True,
    transport: httpx.AsyncBaseTransport | None = None,
) -> httpx.Response:
    """GET with manual redirect following; re-validate every hop (SSRF-safe)."""
    current_url = url
    req_headers = dict(headers) if headers else None
    client_kwargs: dict[str, object] = {
        "timeout": timeout,
        "follow_redirects": False,
        "verify": verify,
    }
    if transport is not None:
        client_kwargs["transport"] = transport

    async with httpx.AsyncClient(**client_kwargs) as client:
        for _ in range(max_redirects):
            assert_public_url(current_url)
            response = await client.get(current_url, headers=req_headers)
            if 300 <= response.status_code < 400:
                location = response.headers.get("location")
                if not location:
                    return response
                current_url = urljoin(current_url, location)
                continue
            return response

    raise ValueError("Too many redirects")

File: lib/ssrfGuard.ts
import dns from 'dns';
import net from 'net';

// SSRF guard for any server-side fetch of a user/DB-derived URL (render-page, image-proxy, favicon,
// competitor/sitemap scraping). Blocks non-http(s) schemes and hosts that resolve to private,
// loopback, link-local (incl. cloud-metadata 169.254.169.254) or reserved ranges.

function ipv4IsPrivate(ip: string): boolean {
  const p = ip.split('.').map(Number);
  if (p.length !== 4 || p.some((n) => Number.isNaN(n) || n < 0 || n > 255)) return true; // malformed → block
  const [a, b] = p;
  if (a === 10) return true;                          // 10.0.0.0/8
  if (a === 127) return true;                         // 127.0.0.0/8 loopback
  if (a === 0) return true;                           // 0.0.0.0/8
  if (a === 169 && b === 254) return true;            // 169.254.0.0/16 link-local (AWS/GCP metadata)
  if (a === 172 && b >= 16 && b <= 31) return true;   // 172.16.0.0/12
  if (a === 192 && b === 168) return true;            // 192.168.0.0/16
  if (a === 100 && b >= 64 && b <= 127) return true;  // 100.64.0.0/10 CGNAT
  if (a >= 224) return true;                          // multicast / reserved
  return false;
}

function ipv6IsPrivate(ip: string): boolean {
  const v = ip.toLowerCase().replace(/^\[|\]$/g, '');
  if (v === '::1' || v === '::') return true;                 // loopback / unspecified
  if (v.startsWith('fc') || v.startsWith('fd')) return true;  // fc00::/7 unique-local
  if (v.startsWith('fe80')) return true;                      // fe80::/10 link-local
  const mapped = v.match(/::ffff:(\d+\.\d+\.\d+\.\d+)$/);     // IPv4-mapped (::ffff:127.0.0.1)
  if (mapped) return ipv4IsPrivate(mapped[1]);
  return false;
}

export function isPrivateAddress(ip: string): boolean {
  const kind = net.isIP(ip);
  if (kind === 4) return ipv4IsPrivate(ip);
  if (kind === 6) return ipv6IsPrivate(ip);
  return true; // not a literal IP we recognize → block to be safe
}

/**
 * Validate that `rawUrl` is an http(s) URL whose host does NOT resolve to a private/reserved
 * address. Throws on any violation (caller returns 400). Returns the parsed URL on success.
 * NOTE: this validates the initial host only — redirects must be disabled or re-validated by
 * the caller (fetch `redirect: 'manual'`, puppeteer redirect interception).
 */
export async function assertPublicUrl(rawUrl: string): Promise<URL> {
  let u: URL;
  try { u = new URL(rawUrl); } catch { throw new Error('Invalid URL'); }
  if (u.protocol !== 'http:' && u.protocol !== 'https:') throw new Error('Only http(s) URLs are allowed');
  const host = u.hostname.replace(/^\[|\]$/g, '');
  if (/^(localhost|.*\.localhost)$/i.test(host)) throw new Error('Blocked host');
  // Literal IP in the URL → check directly (no DNS).
  if (net.isIP(host)) {
    if (isPrivateAddress(host)) throw new Error('Blocked private address');
    return u;
  }
  // Hostname → resolve ALL addresses and reject if any is private (DNS-rebinding-resistant).
  let addrs: { address: string }[];
  try { addrs = await dns.promises.lookup(host, { all: true }); } catch { throw new Error('DNS resolution failed'); }
  if (!addrs.length) throw new Error('DNS resolution failed');
  for (const a of addrs) if (isPrivateAddress(a.address)) throw new Error('Blocked private address');
  return u;
}

/**
 * Fetch following redirects manually, re-validating every hop with assertPublicUrl.
 * Use this instead of fetch(..., { redirect: 'follow' }) / axios maxRedirects for user URLs.
 */
export async function ssrfSafeFetch(
  initial: string,
  init: RequestInit = {},
  maxRedirects = 4,
): Promise<Response> {
  let current = initial;
  for (let i = 0; i < maxRedirects; i += 1) {
    await assertPublicUrl(current);
    // eslint-disable-next-line no-await-in-loop
    const r = await fetch(current, { ...init, redirect: 'manual' });
    if (r.status >= 300 && r.status < 400) {
      const loc = r.headers.get('location');
      if (!loc) return r;
      current = new URL(loc, current).toString();
      continue;
    }
    return r;
  }
  throw new Error('Too many redirects');
}
